### ISSUE TITLE
loss: nan

### ISSUE THREAD

**Issue by lszp** [salesforce/densecap#23]
Hi,
When running the scripts/train.py on three XP GPU  cards, I met  that loss is nan in 1 epoch, likes as follow:

iter: [612/801], training loss: 1.8282, class: 0.1796, reg: 0.1497, sentence: 0.8524, mask: 0.6464, grad norm: 0.6423 data time: 0.0511s, total time: 1.4949s
iter: [613/801], training loss: 1.8322, class: 0.2033, reg: 0.1337, sentence: 0.8701, mask: 0.6250, grad norm: 0.5343 data time: 0.0889s, total time: 1.5393s
iter: [614/801], training loss: 1.8656, class: 0.1751, reg: 0.1262, sentence: 0.9129, mask: 0.6513, grad norm: 0.4983 data time: 1.2910s, total time: 2.6527s
iter: [615/801], training loss: 1.7889, class: 0.2391, reg: 0.1142, sentence: 0.7949, mask: 0.6407, grad norm: 0.5536 data time: 0.0579s, total time: 1.4536s
iter: [616/801], training loss: 1.7299, class: 0.1802, reg: 0.1308, sentence: 0.7956, mask: 0.6232, grad norm: 0.5174 data time: 0.0596s, total time: 1.4421s
iter: [617/801], training loss: 1.8347, class: 0.2067, reg: 0.1366, sentence: 0.8441, mask: 0.6473, grad norm: nan data time: 0.6184s, total time: 2.0576s
iter: [618/801], training loss: nan, class: nan, reg: 0.1438, sentence: nan, mask: 0.6356, grad norm: nan data time: 0.0540s, total time: 1.5435s
iter: [619/801], training loss: nan, class: nan, reg: nan, sentence: nan, mask: nan, grad norm: nan data time: 0.0552s, total time: 1.4475s
iter: [620/801], training loss: nan, class: nan, reg: nan, sentence: nan, mask: nan, grad norm: nan data time: 0.5287s, total time: 2.5449s
iter: [621/801], training loss: nan, class: nan, reg: nan, sentence: nan, mask: nan, grad norm: nan data time: 0.0625s, total time: 1.5136s
iter: [622/801], training loss: nan, class: nan, reg: nan, sentence: nan, mask: nan, grad norm: nan data time: 0.0550s, total time: 1.4975s
iter: [623/801], training loss: nan, class: nan, reg: nan, sentence: nan, mask: nan, grad norm: nan data time: 0.0587s, total time: 1.4647s
iter: [624/801], training loss: nan, class: nan, reg: nan, sentence: nan, mask: nan, grad norm: nan data time: 0.0507s, total time: 1.4093s
iter: [625/801], training loss: nan, class: nan, reg: nan, sentence: nan, mask: nan, grad norm: nan data time: 0.0595s, total time: 1.4266s

Could you tell me that  what may result in it ?

**Comment by LuoweiZhou** [comment]
There are two possibilities:
i) The torch.exp [here](https://github.com/salesforce/densecap/blob/master/model/action_prop_dense_cap.py#L203) and the cross-entropy [loss](https://github.com/salesforce/densecap/blob/master/model/action_prop_dense_cap.py#L295) on the mask cause the numerical unstableness. The trick [here](http://cs231n.github.io/linear-classify/) (see below) could potentially solve the issue. Besides, this only happens when we learn the masks, but not with GT masks.
![image](https://user-images.githubusercontent.com/14948792/67151912-43089780-f29b-11e9-909f-28640b5ba1bb.png)
ii) Depending on how large is your batch size (I assume smaller than the one we have), you will need to reduce the lr.

**Comment by lszp** [comment]
@LuoweiZhou , thank you, my batch size is 45, it works when reducing the lr to 0.01 .

**Comment by LuoweiZhou** [comment]
How about around `45/56*0.1=0.08`?

**Comment by lszp** [comment]
In the first three epochs, the loss is ok when the lr is set to 0.08 .

**Comment by LuoweiZhou** [comment]
You mean you got `nan` after the first three epochs?

**Comment by lszp** [comment]
No. At that time , I only run three epochs. Now it is ok when running all epochs.

**Comment by awkrail** [comment]
I met the same problem, and I found that loss becomes nan when std is zero at LayerNorm in Transformer. 
This is related to the PyTorch issue https://github.com/pytorch/pytorch/issues/4320, and fixed at the latest version.
To avoid this,  instead of `torch.std`, I implement LayerNorm as:
```python
def forward(self, x):
     u = x.mean(-1, keepdim=True)
     s = (x - u).pow(2).mean(-1, keepdim=True)
     x = (x - u) / torch.sqrt(s + self.eps) # prevent std == 0
     return self.gamma * x  + self.beta
```
